diff --git a/src/managers/KeybindManager.cpp b/src/managers/KeybindManager.cpp
index c9c512ae..d91d4471 100644
--- a/src/managers/KeybindManager.cpp
+++ b/src/managers/KeybindManager.cpp
@@ -2753,12 +2753,15 @@ void CKeybindManager::moveWindowOutOfGroup(PHLWINDOW pWindow, const std::string&
     WP<Desktop::View::CGroup> group = pWindow->m_group;
 
     const auto                direction = !dir.empty() ? Math::fromChar(dir[0]) : Math::DIRECTION_DEFAULT;
 
     pWindow->m_group->remove(pWindow, direction);
 
+    if (pWindow->m_groupRules & Desktop::View::GROUP_SET_ALWAYS)
+        Desktop::View::CGroup::create({pWindow});
+
     if (*BFOCUSREMOVEDWINDOW || !group) {
         Desktop::focusState()->fullWindowFocus(pWindow, Desktop::FOCUS_REASON_KEYBIND);
         pWindow->warpCursor();
     } else {
         Desktop::focusState()->fullWindowFocus(group->current(), Desktop::FOCUS_REASON_KEYBIND);
         group->current()->warpCursor();
