diff --git a/niri-config/src/layer_rule.rs b/niri-config/src/layer_rule.rs
index cf099b3a..26f32017 100644
--- a/niri-config/src/layer_rule.rs
+++ b/niri-config/src/layer_rule.rs
@@ -9,12 +9,14 @@ pub struct LayerRule {
     #[knuffel(children(name = "exclude"))]
     pub excludes: Vec<Match>,
 
     #[knuffel(child, unwrap(argument))]
     pub opacity: Option<f32>,
     #[knuffel(child, unwrap(argument))]
+    pub priority: Option<i32>,
+    #[knuffel(child, unwrap(argument))]
     pub block_out_from: Option<BlockOutFrom>,
     #[knuffel(child, default)]
     pub shadow: ShadowRule,
     #[knuffel(child)]
     pub geometry_corner_radius: Option<CornerRadius>,
     #[knuffel(child, unwrap(argument))]
diff --git a/niri-config/src/lib.rs b/niri-config/src/lib.rs
index 329b7669..38671aae 100644
--- a/niri-config/src/lib.rs
+++ b/niri-config/src/lib.rs
@@ -904,12 +904,13 @@ mod tests {
                     active-color "#f00"
                 }
             }
 
             layer-rule {
                 match namespace="^notifications$"
+                priority 1
                 block-out-from "screencast"
             }
 
             binds {
                 Mod+Escape hotkey-overlay-title="Inhibit" { toggle-keyboard-shortcuts-inhibit; }
                 Mod+Shift+Escape allow-inhibiting=true { toggle-keyboard-shortcuts-inhibit; }
@@ -1904,12 +1905,15 @@ mod tests {
                             at_startup: None,
                             layer: None,
                         },
                     ],
                     excludes: [],
                     opacity: None,
+                    priority: Some(
+                        1,
+                    ),
                     block_out_from: Some(
                         Screencast,
                     ),
                     shadow: ShadowRule {
                         off: false,
                         on: false,
diff --git a/src/layer/mod.rs b/src/layer/mod.rs
index 767a6b0e..4703a527 100644
--- a/src/layer/mod.rs
+++ b/src/layer/mod.rs
@@ -10,12 +10,14 @@ pub use mapped::MappedLayer;
 /// Rules fully resolved for a layer-shell surface.
 #[derive(Debug, Default, PartialEq)]
 pub struct ResolvedLayerRules {
     /// Extra opacity to draw this layer surface with.
     pub opacity: Option<f32>,
 
+    pub priority: i32,
+
     /// Whether to block out this layer surface from certain render targets.
     pub block_out_from: Option<BlockOutFrom>,
 
     /// Shadow overrides.
     pub shadow: ShadowRule,
 
@@ -60,12 +62,15 @@ impl ResolvedLayerRules {
                 continue;
             }
 
             if let Some(x) = rule.opacity {
                 resolved.opacity = Some(x);
             }
+            if let Some(x) = rule.priority {
+                resolved.priority = x;
+            }
             if let Some(x) = rule.block_out_from {
                 resolved.block_out_from = Some(x);
             }
             if let Some(x) = rule.geometry_corner_radius {
                 resolved.geometry_corner_radius = Some(x);
             }
diff --git a/src/niri.rs b/src/niri.rs
index ec40e01c..bc006793 100644
--- a/src/niri.rs
+++ b/src/niri.rs
@@ -4524,22 +4524,27 @@ impl Niri {
         &'a self,
         layer_map: &'a LayerMap,
         layer: Layer,
         for_backdrop: bool,
     ) -> impl Iterator<Item = (&'a MappedLayer, Rectangle<i32, Logical>)> {
         // LayerMap returns layers in reverse stacking order.
-        layer_map.layers_on(layer).rev().filter_map(move |surface| {
-            let mapped = self.mapped_layer_surfaces.get(surface)?;
+        let mut layers = layer_map
+            .layers_on(layer)
+            .filter_map(move |surface| {
+                let mapped = self.mapped_layer_surfaces.get(surface)?;
 
-            if for_backdrop != mapped.place_within_backdrop() {
-                return None;
-            }
+                if for_backdrop != mapped.place_within_backdrop() {
+                    return None;
+                }
 
-            let geo = layer_map.layer_geometry(surface)?;
-            Some((mapped, geo))
-        })
+                let geo = layer_map.layer_geometry(surface)?;
+                Some((mapped, geo))
+            })
+            .collect::<Vec<_>>();
+        layers.sort_by_key(|(mapped, _)| mapped.rules().priority);
+        layers.into_iter().rev()
     }
 
     #[allow(clippy::too_many_arguments)]
     fn render_layer_normal<R: NiriRenderer>(
         &self,
         mut ctx: RenderCtx<R>,
