diff --git a/src/niri.rs b/src/niri.rs
index c65b9d1d..aaf6c108 100644
--- a/src/niri.rs
+++ b/src/niri.rs
@@ -5000,13 +5000,13 @@ impl Niri {
             .or_default();
 
         // Reset the delay state.
         force_render_state.last_render_time = frame_callback_time;
         force_render_state.is_waiting_for_render = false;
 
-        debug!("Sending frame callback for surface");
+        trace!("Sending frame callback for surface");
 
         send_frames_surface_tree(
             &surface,
             output,
             frame_callback_time,
             FRAME_CALLBACK_THROTTLE,
@@ -5100,13 +5100,13 @@ impl Niri {
                                 .entry(surface.id())
                                 .or_insert((surface.clone(), delay_time));
 
                             // Set force render state.
                             force_render_state.is_waiting_for_render = true;
                         }
-                        return None;
+                        None
                     } else {
                         // Remove force render state if not on primary output.
                         if on_primary {
                             force_render_state_borrow.remove(&surface.id().protocol_id());
                         }
                         //Fps unlimited branch.
@@ -5188,12 +5188,13 @@ impl Niri {
                 FRAME_CALLBACK_THROTTLE,
                 should_send,
             );
         }
 
         // Call delay timer
+        #[allow(clippy::mutable_key_type)]
         let delayed_map = delayed_surfaces.into_inner();
         for (_, (surface, interval)) in delayed_map {
             self.delay_to_send_frame_callbacks(surface, output.clone(), interval);
         }
     }
 
