diff --git a/niri-config/src/layout.rs b/niri-config/src/layout.rs
index f84528bd..f8b2757c 100644
--- a/niri-config/src/layout.rs
+++ b/niri-config/src/layout.rs
@@ -12,13 +12,13 @@ pub struct Layout {
     pub focus_ring: FocusRing,
     pub border: Border,
     pub shadow: Shadow,
     pub tab_indicator: TabIndicator,
     pub insert_hint: InsertHint,
     pub preset_column_widths: Vec<PresetSize>,
-    pub default_column_width: Option<PresetSize>,
+    pub default_column_width: DefaultColumnWidth,
     pub preset_window_heights: Vec<PresetSize>,
     pub center_focused_column: CenterFocusedColumn,
     pub always_center_single_column: bool,
     pub empty_workspace_above_first: bool,
     pub default_column_display: ColumnDisplay,
     pub gaps: f64,
@@ -36,13 +36,16 @@ impl Default for Layout {
             insert_hint: InsertHint::default(),
             preset_column_widths: vec![
                 PresetSize::Proportion(1. / 3.),
                 PresetSize::Proportion(0.5),
                 PresetSize::Proportion(2. / 3.),
             ],
-            default_column_width: Some(PresetSize::Proportion(0.5)),
+            default_column_width: DefaultColumnWidth {
+                size: Some(PresetSize::Proportion(0.5)),
+                maximize: false,
+            },
             center_focused_column: CenterFocusedColumn::Never,
             always_center_single_column: false,
             empty_workspace_above_first: false,
             default_column_display: ColumnDisplay::Normal,
             gaps: 16.,
             struts: Struts::default(),
@@ -78,13 +81,13 @@ impl MergeWith<LayoutPart> for Layout {
             default_column_display,
             struts,
             background_color,
         );
 
         if let Some(x) = part.default_column_width {
-            self.default_column_width = x.0;
+            self.default_column_width = x;
         }
 
         if self.preset_column_widths.is_empty() {
             self.preset_column_widths = Layout::default().preset_column_widths;
         }
 
@@ -106,13 +109,13 @@ pub struct LayoutPart {
     pub tab_indicator: Option<TabIndicatorPart>,
     #[knuffel(child)]
     pub insert_hint: Option<InsertHintPart>,
     #[knuffel(child, unwrap(children))]
     pub preset_column_widths: Option<Vec<PresetSize>>,
     #[knuffel(child)]
-    pub default_column_width: Option<DefaultPresetSize>,
+    pub default_column_width: Option<DefaultColumnWidth>,
     #[knuffel(child, unwrap(children))]
     pub preset_window_heights: Option<Vec<PresetSize>>,
     #[knuffel(child, unwrap(argument))]
     pub center_focused_column: Option<CenterFocusedColumn>,
     #[knuffel(child)]
     pub always_center_single_column: Option<Flag>,
@@ -143,12 +146,18 @@ impl From<PresetSize> for SizeChange {
     }
 }
 
 #[derive(Debug, Clone, Copy, PartialEq)]
 pub struct DefaultPresetSize(pub Option<PresetSize>);
 
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct DefaultColumnWidth {
+    pub size: Option<PresetSize>,
+    pub maximize: bool,
+}
+
 #[derive(knuffel::Decode, Debug, Default, Clone, Copy, PartialEq)]
 pub struct Struts {
     #[knuffel(child, unwrap(argument), default)]
     pub left: FloatOrInt<-65535, 65535>,
     #[knuffel(child, unwrap(argument), default)]
     pub right: FloatOrInt<-65535, 65535>,
@@ -193,6 +202,66 @@ where
             PresetSize::decode_node(child, ctx).map(Some).map(Self)
         } else {
             Ok(Self(None))
         }
     }
 }
+
+impl<S> knuffel::Decode<S> for DefaultColumnWidth
+where
+    S: knuffel::traits::ErrorSpan,
+{
+    fn decode_node(
+        node: &knuffel::ast::SpannedNode<S>,
+        ctx: &mut knuffel::decode::Context<S>,
+    ) -> Result<Self, DecodeError<S>> {
+        if let Some(type_name) = &node.type_name {
+            ctx.emit_error(DecodeError::unexpected(
+                type_name,
+                "type name",
+                "no type name expected for this node",
+            ));
+        }
+
+        for val in node.arguments.iter() {
+            ctx.emit_error(DecodeError::unexpected(
+                &val.literal,
+                "argument",
+                "no arguments expected for this node",
+            ))
+        }
+
+        let mut maximize = false;
+        for (name, val) in &node.properties {
+            match &***name {
+                "maximize" => {
+                    maximize = knuffel::DecodeScalar::decode(val, ctx)?;
+                }
+                name_str => {
+                    ctx.emit_error(DecodeError::unexpected(
+                        name,
+                        "property",
+                        format!("unexpected property`{}`", name_str.escape_default()),
+                    ));
+                }
+            }
+        }
+
+        let size: Option<PresetSize> = {
+            let mut children = node.children();
+            if let Some(child) = children.next() {
+                if let Some(unwanted_child) = children.next() {
+                    ctx.emit_error(DecodeError::unexpected(
+                        unwanted_child,
+                        "node",
+                        "expected no more than one child",
+                    ));
+                }
+                Some(PresetSize::decode_node(child, ctx)?)
+            } else {
+                None
+            }
+        };
+
+        Ok(Self { size, maximize })
+    }
+}
diff --git a/niri-config/src/lib.rs b/niri-config/src/lib.rs
index 909aeb80..b4f837cb 100644
--- a/niri-config/src/lib.rs
+++ b/niri-config/src/lib.rs
@@ -797,13 +797,13 @@ mod tests {
                     proportion 0.25
                     proportion 0.5
                     fixed 960
                     fixed 1280
                 }
 
-                default-column-width { proportion 0.25; }
+                default-column-width maximize=true { proportion 0.25; }
 
                 gaps 8
 
                 struts {
                     left 1
                     right 2
@@ -1426,17 +1426,20 @@ mod tests {
                         960,
                     ),
                     Fixed(
                         1280,
                     ),
                 ],
-                default_column_width: Some(
-                    Proportion(
-                        0.25,
+                default_column_width: DefaultColumnWidth {
+                    size: Some(
+                        Proportion(
+                            0.25,
+                        ),
                     ),
-                ),
+                    maximize: true,
+                },
                 preset_window_heights: [
                     Proportion(
                         0.25,
                     ),
                     Proportion(
                         0.5,
diff --git a/src/handlers/xdg_shell.rs b/src/handlers/xdg_shell.rs
index 38440c90..7962eb3b 100644
--- a/src/handlers/xdg_shell.rs
+++ b/src/handlers/xdg_shell.rs
@@ -1105,13 +1105,14 @@ impl State {
         let mon = mon.map(|(mon, _)| mon);
 
         let mut width = None;
         let mut floating_width = None;
         let mut height = None;
         let mut floating_height = None;
-        let is_full_width = rules.open_maximized.unwrap_or(false);
+        let is_full_width_maybe = rules.open_maximized;
+        let mut is_full_width = is_full_width_maybe.unwrap_or(false);
         let is_floating = rules.compute_open_floating(toplevel);
 
         // Tell the surface the preferred size and bounds for its likely output.
         let ws = rules
             .open_on_workspace
             .as_deref()
@@ -1136,12 +1137,13 @@ impl State {
             } else if is_pending_maximized {
                 toplevel.with_pending_state(|state| {
                     state.states.set(xdg_toplevel::State::Maximized);
                 });
             }
 
+            is_full_width = ws.resolve_default_full_width(is_full_width_maybe);
             width = ws.resolve_default_width(rules.default_width, false);
             floating_width = ws.resolve_default_width(rules.default_width, true);
             height = ws.resolve_default_height(rules.default_height, false);
             floating_height = ws.resolve_default_height(rules.default_height, true);
 
             let configure_width = if is_floating {
diff --git a/src/layout/workspace.rs b/src/layout/workspace.rs
index 74487797..ee1f9640 100644
--- a/src/layout/workspace.rs
+++ b/src/layout/workspace.rs
@@ -784,25 +784,33 @@ impl<W: LayoutElement> Workspace<W> {
 
         self.update_focus_floating_tiling_after_removing(from_floating);
 
         Some(column)
     }
 
+    pub fn resolve_default_full_width(&self, full_width: Option<bool>) -> bool {
+        full_width.unwrap_or(self.options.layout.default_column_width.maximize)
+    }
+
     pub fn resolve_default_width(
         &self,
         default_width: Option<Option<PresetSize>>,
         is_floating: bool,
     ) -> Option<PresetSize> {
         match default_width {
             Some(Some(width)) => Some(width),
             Some(None) => None,
             None if is_floating => None,
-            None => self.options.layout.default_column_width,
+            None => self.options.layout.default_column_width.size,
         }
     }
 
+    pub fn get_default_maximize(&self) -> bool {
+        self.options.layout.default_column_width.maximize
+    }
+
     pub fn resolve_default_height(
         &self,
         default_height: Option<Option<PresetSize>>,
         is_floating: bool,
     ) -> Option<PresetSize> {
         match default_height {
