diff --git a/niri-config/src/lib.rs b/niri-config/src/lib.rs
index 909aeb80..de432b6b 100644
--- a/niri-config/src/lib.rs
+++ b/niri-config/src/lib.rs
@@ -1887,12 +1887,14 @@ mod tests {
                             xray: None,
                             blur: None,
                             noise: None,
                             saturation: None,
                         },
                     },
+                    force_render: None,
+                    force_render_fps: None,
                 },
             ],
             layer_rules: [
                 LayerRule {
                     matches: [
                         Match {
diff --git a/niri-config/src/window_rule.rs b/niri-config/src/window_rule.rs
index f2bc2ad1..543b863f 100644
--- a/niri-config/src/window_rule.rs
+++ b/niri-config/src/window_rule.rs
@@ -76,12 +76,16 @@ pub struct WindowRule {
     #[knuffel(child, unwrap(argument))]
     pub tiled_state: Option<bool>,
     #[knuffel(child, default)]
     pub background_effect: BackgroundEffectRule,
     #[knuffel(child, default)]
     pub popups: PopupsRule,
+    #[knuffel(child, unwrap(argument))]
+    pub force_render: Option<bool>,
+    #[knuffel(child, unwrap(argument))]
+    pub force_render_fps: Option<u16>,
 }
 
 /// Rules for popup surfaces.
 #[derive(knuffel::Decode, Debug, Default, Clone, PartialEq)]
 pub struct PopupsRule {
     #[knuffel(child, unwrap(argument))]
diff --git a/src/niri.rs b/src/niri.rs
index 190ef09d..1c49e9d9 100644
--- a/src/niri.rs
+++ b/src/niri.rs
@@ -411,12 +411,14 @@ pub struct Niri {
     pub ipc_outputs_changed: bool,
 
     pub satellite: Option<Satellite>,
 
     #[cfg(feature = "xdp-gnome-screencast")]
     pub casting: Screencasting,
+
+    pub force_render_state: RefCell<HashMap<u32, ForceRenderState>>,
 }
 
 #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
 pub enum PointerVisibility {
     /// The pointer is visible.
     Visible,
@@ -564,12 +566,28 @@ pub enum LockRenderState {
 // This state keeps track of when a surface last received a frame callback.
 struct SurfaceFrameThrottlingState {
     /// Output and sequence that the frame callback was last sent at.
     last_sent_at: RefCell<Option<(Output, u32)>>,
 }
 
+pub struct ForceRenderState {
+    // To calculate time diff between current frame_callback_time and last render time.
+    last_render_time: Duration,
+    // To check if a force render frame is waiting to be rendered.
+    is_waiting_for_render: bool,
+}
+
+impl Default for ForceRenderState {
+    fn default() -> Self {
+        Self {
+            last_render_time: Duration::ZERO,
+            is_waiting_for_render: false,
+        }
+    }
+}
+
 pub enum CenterCoords {
     Separately,
     Both,
     // Force centering even if the cursor is already in the rectangle.
     BothAlways,
 }
@@ -2622,12 +2640,14 @@ impl Niri {
             ipc_outputs_changed: false,
 
             satellite: None,
 
             #[cfg(feature = "xdp-gnome-screencast")]
             casting: screencasting,
+
+            force_render_state: RefCell::new(HashMap::new()),
         };
 
         niri.reset_pointer_inactivity_timer();
 
         niri
     }
@@ -5009,12 +5029,52 @@ impl Niri {
                     )
                 },
             );
         }
     }
 
+    pub fn delay_to_send_frame_callbacks(
+        &mut self,
+        surface: WlSurface,
+        output: Output,
+        interval: Duration,
+    ) {
+        let timer = Timer::from_duration(interval);
+        self.event_loop
+            .insert_source(timer, move |_, _, state| {
+                state
+                    .niri
+                    .send_frame_callback_for_surface(surface.clone(), &output);
+                TimeoutAction::Drop
+            })
+            .unwrap();
+    }
+
+    fn send_frame_callback_for_surface(&mut self, surface: WlSurface, output: &Output) {
+        let frame_callback_time = get_monotonic_time();
+        let mut force_render_state_borrow = self.force_render_state.borrow_mut();
+
+        let force_render_state = force_render_state_borrow
+            .entry(surface.id().protocol_id())
+            .or_default();
+
+        // Reset the delay state.
+        force_render_state.last_render_time = frame_callback_time;
+        force_render_state.is_waiting_for_render = false;
+
+        trace!("Sending frame callback for surface");
+
+        send_frames_surface_tree(
+            &surface,
+            output,
+            frame_callback_time,
+            FRAME_CALLBACK_THROTTLE,
+            |_, _| Some(output.clone()),
+        );
+    }
+
     pub fn send_frame_callbacks(&mut self, output: &Output) {
         let _span = tracy_client::span!("Niri::send_frame_callbacks");
 
         let state = self.output_state.get(output).unwrap();
         let sequence = state.frame_callback_sequence;
 
@@ -5050,19 +5110,107 @@ impl Niri {
                 None
             }
         };
 
         let frame_callback_time = get_monotonic_time();
 
+        // Collect delayed surfaces.
+        let delayed_surfaces = RefCell::new(HashMap::new());
+
         for mapped in self.layout.windows_for_output_mut(output) {
-            mapped.send_frame(
-                output,
-                frame_callback_time,
-                FRAME_CALLBACK_THROTTLE,
-                should_send,
-            );
+            // Check if the surface should be forced to render.
+            if mapped.rules().force_render == Some(true) || mapped.is_window_cast_target() {
+                // Calculate delay time.
+                let interval = if let Some(force_render_fps) = mapped.rules().force_render_fps {
+                    if force_render_fps == 0 {
+                        Duration::ZERO
+                    } else {
+                        Duration::from_secs_f64(1.0 / force_render_fps as f64)
+                    }
+                } else {
+                    Duration::ZERO
+                };
+
+                let should_force_render = |surface: &WlSurface, states: &SurfaceData| {
+                    // Check if the surface is on the primary output.
+                    let mut on_primary = true;
+                    let current_primary_output = surface_primary_scanout_output(surface, states);
+                    if current_primary_output.as_ref() != Some(output) {
+                        on_primary = false;
+                    }
+
+                    let mut force_render_state_borrow = self.force_render_state.borrow_mut();
+
+                    if !on_primary && !interval.is_zero() {
+                        //Fps limited branch.
+                        let force_render_state = force_render_state_borrow
+                            .entry(surface.id().protocol_id())
+                            .or_default();
+
+                        if !force_render_state.is_waiting_for_render {
+                            // Time diff since last frame callback.
+                            let time_diff: Duration = frame_callback_time
+                                .saturating_sub(force_render_state.last_render_time);
+
+                            // Calculate next frame callback time and push needed data to
+                            // delayed_surfaces.
+                            let delay_time = interval.saturating_sub(time_diff);
+                            let mut delayed = delayed_surfaces.borrow_mut();
+                            delayed
+                                .entry(surface.id())
+                                .or_insert((surface.clone(), delay_time));
+
+                            // Set force render state.
+                            force_render_state.is_waiting_for_render = true;
+                        }
+                        None
+                    } else {
+                        // Remove force render state if not on primary output.
+                        if on_primary {
+                            force_render_state_borrow.remove(&surface.id().protocol_id());
+                        }
+                        //Fps unlimited branch.
+                        let frame_throttling_state = states
+                            .data_map
+                            .get_or_insert(SurfaceFrameThrottlingState::default);
+
+                        // Next, check the throttling status.
+                        let mut last_sent_at = frame_throttling_state.last_sent_at.borrow_mut();
+                        let mut send = true;
+
+                        // If we already sent a frame callback to this surface this output refresh
+                        // cycle, don't send one again to prevent empty-damage commit busy loops.
+                        if let Some((last_output, last_sequence)) = &*last_sent_at {
+                            if last_output == output && *last_sequence == sequence {
+                                send = false;
+                            }
+                        }
+
+                        if send {
+                            *last_sent_at = Some((output.clone(), sequence));
+                            Some(output.clone())
+                        } else {
+                            None
+                        }
+                    }
+                };
+
+                mapped.send_frame(
+                    output,
+                    frame_callback_time,
+                    FRAME_CALLBACK_THROTTLE,
+                    should_force_render,
+                );
+            } else {
+                mapped.send_frame(
+                    output,
+                    frame_callback_time,
+                    FRAME_CALLBACK_THROTTLE,
+                    should_send,
+                );
+            }
         }
 
         for surface in layer_map_for_output(output).layers() {
             surface.send_frame(
                 output,
                 frame_callback_time,
@@ -5097,12 +5245,19 @@ impl Niri {
                 output,
                 frame_callback_time,
                 FRAME_CALLBACK_THROTTLE,
                 should_send,
             );
         }
+
+        // Call delay timer
+        #[allow(clippy::mutable_key_type)]
+        let delayed_map = delayed_surfaces.into_inner();
+        for (_, (surface, interval)) in delayed_map {
+            self.delay_to_send_frame_callbacks(surface, output.clone(), interval);
+        }
     }
 
     pub fn send_frame_callbacks_on_fallback_timer(&mut self) {
         let _span = tracy_client::span!("Niri::send_frame_callbacks_on_fallback_timer");
 
         // Make up a bogus output; we don't care about it here anyway, just the throttling timer.
diff --git a/src/window/mod.rs b/src/window/mod.rs
index 14527cd2..adfc140c 100644
--- a/src/window/mod.rs
+++ b/src/window/mod.rs
@@ -122,12 +122,18 @@ pub struct ResolvedWindowRules {
 
     /// Background effect configuration.
     pub background_effect: BackgroundEffect,
 
     /// Rules for this window's popups.
     pub popups: ResolvedPopupsRules,
+
+    /// Whether to force a render on this window.
+    pub force_render: Option<bool>,
+
+    /// force render fps limit for this window.
+    pub force_render_fps: Option<u16>,
 }
 
 impl<'a> WindowRef<'a> {
     pub fn toplevel(self) -> &'a ToplevelSurface {
         match self {
             WindowRef::Unmapped(unmapped) => unmapped.toplevel(),
@@ -300,12 +306,20 @@ impl ResolvedWindowRules {
                     resolved.scroll_factor = Some(x.0);
                 }
                 if let Some(x) = rule.tiled_state {
                     resolved.tiled_state = Some(x);
                 }
 
+                if let Some(x) = rule.force_render {
+                    resolved.force_render = Some(x);
+                }
+
+                if let Some(x) = rule.force_render_fps {
+                    resolved.force_render_fps = Some(x);
+                }
+
                 resolved
                     .background_effect
                     .merge_with(&rule.background_effect);
 
                 resolved.popups.merge_with(&rule.popups);
             }
