diff --git a/src/alejandra_cli/src/cli.rs b/src/alejandra_cli/src/cli.rs
index 0ef0496..97860bc 100644
--- a/src/alejandra_cli/src/cli.rs
+++ b/src/alejandra_cli/src/cli.rs
@@ -6,13 +6,12 @@ use clap::value_parser;
 use clap::ArgAction;
 use clap::Parser;
 use futures::future::RemoteHandle;
 use futures::stream::FuturesUnordered;
 use futures::task::SpawnExt;
 
-use crate::ads::random_ad;
 use crate::verbosity::Verbosity;
 
 /// The Uncompromising Nix Code Formatter.
 #[derive(Debug, Parser)]
 #[clap(
     name="Alejandra",
@@ -101,13 +100,12 @@ fn format_paths(
     if verbosity.allows_info() {
         eprintln!(
             "Checking style in {paths_len} file{} using {threads} thread{}.",
             if paths_len == 1 { "" } else { "s" },
             if threads == 1 { "" } else { "s" },
         );
-        eprintln!();
     }
 
     let pool = futures::executor::ThreadPoolBuilder::new()
         .pool_size(threads)
         .create()
         .expect("Unable to instantiate a new thread pool.");
@@ -206,43 +204,28 @@ pub fn main() -> ! {
             _ => false,
         })
         .count();
 
     if changed > 0 {
         if verbosity.allows_info() {
-            eprintln!();
             eprintln!(
                 "{}! {changed} file{} {}.",
                 if in_place { "Success" } else { "Alert" },
                 if changed == 1 { "" } else { "s" },
                 match (changed == 1, in_place) {
                     (false, true) => "were formatted",
                     (false, false) => "require formatting",
                     (true, true) => "was formatted",
                     (true, false) => "requires formatting",
                 }
             );
-
-            if in_place {
-                eprintln!();
-                eprint!("{}", random_ad());
-            }
         }
 
         std::process::exit(if in_place { 0 } else { 2 });
     }
 
-    if verbosity.allows_info() {
-        eprintln!();
-        eprintln!(
-            "Congratulations! Your code complies with the Alejandra style."
-        );
-        eprintln!();
-        eprint!("{}", random_ad());
-    }
-
     std::process::exit(0);
 }
 
 fn resolve_config(path: Option<&str>, verbosity: Verbosity) -> Config {
     let default_config_path = "alejandra.toml";
 
diff --git a/src/alejandra_cli/src/lib.rs b/src/alejandra_cli/src/lib.rs
index fd49ce8..258f656 100644
--- a/src/alejandra_cli/src/lib.rs
+++ b/src/alejandra_cli/src/lib.rs
@@ -1,4 +1,3 @@
-mod ads;
 pub mod cli;
 mod find;
 mod verbosity;
